consensus/istanbul/backend/announce.go

```diff
@@ -497,32 +523,66 @@ func (sb *Backend) handleAnnounceMsg(peer consensus.Peer, payload []byte) error
 		logger.Trace("Going to process an announce msg", "announce records", announceData.AnnounceRecords)
 		for _, announceRecord := range announceData.AnnounceRecords {
 			if announceRecord.DestAddress == sb.Address() {
-				// TODO: Decrypt the enodeURL using this validator's validator key after making changes to encrypt it
-				enodeUrl := string(announceRecord.EncryptedEnodeURL)
-				node, err := enode.ParseV4(enodeUrl)
+				enodeBytes, err := sb.decryptFn(accounts.Account{Address: sb.Address()}, announceRecord.EncryptedEnodeURL, nil, nil)
 				if err != nil {
-					logger.Error("Error in parsing enodeURL", "enodeUrl", enodeUrl)
+					sb.logger.Warn("Error decrypting endpoint", "err", err, "announceRecord.EncryptedEnodeURL", announceRecord.EncryptedEnodeURL)
 					return err
 				}
-
-				if err := sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Version: announceData.Version}}); err != nil {
-					logger.Warn("Error in upserting a valenode entry", "AnnounceData", announceData.String(), "error", err)
+				enodeURL := string(enodeBytes)
+				node, err := enode.ParseV4(enodeURL)
+				if err != nil {
+					logger.Error("Error parsing enodeURL", "enodeUrl", enodeURL)
 					return err
 				}
-
+				if err := sb.answerAnnounceMsg(msg.Address, node, announceData.Version); err != nil {
+					logger.Warn("Error answering an announce msg", "target node", node.URLv4(), "error", err)
+					return err
+				}
+				// If the announce was only intended for this node, do not regossip
+				if len(announceData.AnnounceRecords) == 1 {
+					return nil
+				}
 				break
 			}
 		}
 	}
 
-	sb.cachedAnnounceMsgsMu.Lock()
-	sb.cachedAnnounceMsgs[msg.Address] = &announceMsgCachedEntry{MsgVersion: announceData.Version, MsgPayload: payload}
-	sb.cachedAnnounceMsgsMu.Unlock()
-
 	// Regossip this announce message
 	return sb.regossipAnnounce(msg, payload)
 }
 
+// answerAnnounceMsg will answer a received announce message. If the target
+// node is already a peer of any kind, an enodeCertificate will be sent.
+// Regardless, the target node will be upserted into the val enode table
+// to ensure this node designates the target node as a ValidatorPurpose peer
+// with a valid version.
+func (sb *Backend) answerAnnounceMsg(address common.Address, node *enode.Node, version uint) error {
+	targetID := node.ID()
+	targetIDs := map[enode.ID]bool{
+		targetID: true,
+	}
+	// The target could be an existing peer of any purpose.
+	matches := sb.broadcaster.FindPeers(targetIDs, p2p.AnyPurpose)
+	if matches[targetID] != nil {
+		enodeCertificateMsg, err := sb.retrieveEnodeCertificateMsg()
+		if err != nil {
+			return err
+		}
+		if err := sb.sendEnodeCertificateMsg(matches[targetID], enodeCertificateMsg); err != nil {
```

Comment:
If I understand correctly, we could receive the same gossiped announce message from multiple peers. If so, we should make sure we avoid sending that peer duplicate enode certificates

Reply:
Great point. I'll make use of the cooldown period. I don't think it's safe to "answer" the announce msg and then always refuse to answer any later requests in the event that the answer is received by the querying validator (or needs an answer again in the future for some reason)